### DISCUSSION TITLE
Is there a way to ssh into the VM?

### DISCUSSION THREAD

**Discussion by threefcata** [utmapp/UTM#2465] [UTM Help (macOS)]
**Describe the issue**
I don't wanna fiddle with the GUI inside the VM as i usually ssh into the VM. With QEMU, it usually requires some kind of network bridge on the host, then the VM is launched using that bridge as the underlying network device. Could you please advise how to do that on an M1 macbook? Thanks.

**Configuration**
* UTM Version: 2.0.26
* OS Version: 11.2.3
* Intel or Apple Silicon? Apple

**Comment by osy** [comment]
Duplicate of #2280

**Comment by osy** [comment]
You can use port forwarding but it still requires using the GUI.

**Comment by agreatfool** [comment]
@osy How to do this? Is there any guide? I think this should be a common question, since now there are MAC users using UTM vm.

**Reply by jamespo** [reply]
https://www.webscalability.com/blog/2021/05/access-your-utm-vm-over-ssh/

**Reply by ghost** [reply]
Great! What password you need to submit at 'user@localhost's password' in the Mac terminal? I try my local account and the password of my debian user. I get on both 'Permission denied, please try again.'

**Reply by jamespo** [reply]
Should be the username of your debian user and the password of your debian user.

If the debian username is different then specify it with `ssh debianusername@localhost`

But inside the GUI of your VM you may to want to test if ssh is working first with a simple `ssh localhost`

**Reply by ghost** [reply]
Thanks! It works with `ssh -p 22022 mydebianusername@localhost`

**Reply by zigang93** [reply]
the latest utm interface is different with that blog.. couldn't find any port settings under network

**Reply by Danendra10** [reply]
Try to change network mode to emulated vlan

**Reply by siaulich** [reply]
Hey, I cannot access the internet from within the VM, if I am running on emulated VLAN. Any hints on that?

**Comment by lizrice** [comment]
I had this working great using port-forwarding to ssh into Ubuntu VMs running in UTM ([as described here](https://medium.com/@lizrice/linux-vms-on-an-m1-based-mac-with-vscode-and-utm-d73e7cb06133)) - but I'm actually here because this has stopped working! I have just updated Big Sur to 11.4 and also took an update to UTM Version 2.1.1 (30) since the last time I used it, so I am not sure which of these updates has caused it to stop working 😢

Port forwarding is set up to forward 2200 on the host to 22 on the guest. Debug output from ssh on the host is now

```
❯ ssh 127.0.0.1 -p 2200 -v
OpenSSH_8.1p1, LibreSSL 2.7.3
debug1: Reading configuration data /Users/liz/.ssh/config
debug1: /Users/liz/.ssh/config line 1: Applying options for 127.0.0.1
debug1: /Users/liz/.ssh/config line 7: Applying options for *
debug1: Reading configuration data /etc/ssh/ssh_config
debug1: /etc/ssh/ssh_config line 47: Applying options for *
debug1: Connecting to 127.0.0.1 [127.0.0.1] port 2200.
debug1: Connection established.
debug1: identity file /Users/liz/.ssh/id_rsa type 0
debug1: identity file /Users/liz/.ssh/id_rsa-cert type -1
debug1: Local version string SSH-2.0-OpenSSH_8.1
kex_exchange_identification: read: Connection reset by peer
```
Any assistance or pointers gratefully received

**Reply by ghost** [reply]
Hi! I don't using it with VSCode, but with my Terminal. I use `ssh -p 22022 YourVMusername@localhost` to obtain access.

In your case, try with port 2200.

**Reply by ghost** [reply]
I also checked it with VScode and I can connect on my debian VM with `ssh -p 22022 debian@localhost` using the Remote Development pack;

https://marketplace.visualstudio.com/items?itemName=ms-vscode-remote.vscode-remote-extensionpack

**Reply by jamespo** [reply]
@lizrice can you ssh locally on the actual guest OS itself (just ssh localhost inside Ubuntu)?

**Reply by lizrice** [reply]
@jamespo yes, ssh-ing from the guest to itself works fine

@r1card092 I think that's equivalent to what I posted (because localhost == 127.0.0.1 and my user name's the same on both host & guest, but just to be sure I tried it and the result is the same (this time no debugging with -v, just the result) 

```
❯ ssh -p 2200 liz@localhost
kex_exchange_identification: Connection closed by remote host
```

FWIW I can ssh (and use VSCode remote) just fine from my mac to a cloud-hosted VM. [This](https://serverfault.com/questions/1015547/what-causes-ssh-error-kex-exchange-identification-connection-closed-by-remote) makes me think somehow the port forwarding is not working correctly (although I've changed nothing in my config)

**Reply by jamespo** [reply]
@lizrice I would doublecheck for a local firewall on the guest (unlikely).

Maybe remove the port forwarding & set it up again.

I'm on 11.4 but unsure what version of UTM I'm on

**Reply by ghost** [reply]
@lizrice That was not what I meaning, I hope we all know that 127.0.0.1 is localhost =] I was pointing on your username in host and guest OS.

I would advice download a fresh copy of debian minimal on the UTM gallery page and test the behavior. If it works, there is something in your guest OS we can troubleshoot on.

-

**Reply by lizrice** [reply]
Thanks @r1card092, good idea. Works perfectly with a new debian minimal (`ssh -p 2202 debian@127.0.0.1`) 

I tried removing and re-creating the port forwarding for my Ubuntu VM as per @jamespo's suggestion but that didn't help. Going to try creating a new Ubuntu VM and see how that goes.

**Comment by Inc0n** [comment]
Is the more recent versions, utm: 3.0.4-2
There isn't a option for port forwarding, how should this be configured accordingly?

**Reply by Inc0n** [reply]
Ok nvm, with reference to PR https://github.com/utmapp/UTM/pull/3025, you would need to change Network mode to `Emulated VLAN`

**Reply by francesco-plt** [reply]
If I do that the VM is not able to connect to the internet anymore. Ubuntu server ARM 20.04, Macbook Pro M1.

**Reply by flyx** [reply]
If you set up the VM to `Shared Network` (which seems to be the default as of 3.2.4), you can reach the VM via its ethernet IP.

In the VM, do `ip -4 addr` (`-6` for IPv6), you'll find the IP under your network interface (`enp0s7` for me) behind `inet`. Then on the host, `ssh` to this address.

**Reply by msharran** [reply]
Additional Tip! If you don't have a display device configured, log in via Serial device > Built-In-Terminal and run the above mentioned commands. Then change it back to Pseudo-TTY and ssh via the guest ip on port 22

**Comment by aaidenduaban** [comment]
a great explanation here https://www.youtube.com/watch?v=GgDEwQXpZI8&ab_channel=ryd3v by Ry about UTM networking, and also encountered an error on SSL connection from host to VM. Can connect to VM1 to VM2 and vice versa, VMs to Host. Also tried minimal 
(ssh -p 2202 debian@127.0.0.1) as stated above but still no luck in connecting Host to VM. 
Specs:
(VM) Network: Bridged (Advanced) 
UTM Version: 4.1.5
Ubuntu 22.04
Macbook Pro M1
any help is greatly appreciated

**Reply by ktprograms** [reply]
With Bridged networking, you don't use port forwarding,  so ssh localhost:2202 is wrong.

Try getting the IP from inside the VM, and make sure it's a valid IP that's on the same subnet as the host (the first 3 octets should be the same as your host)

**Reply by aaidenduaban** [reply]
I got my IP from VM but still no luck. Also remove the port and still can't ssh or even ping from host to VM.

**Reply by ktprograms** [reply]
What is the IP?
(Make sure it's the network local IP and not your public IP)

**Reply by TnGadaria** [reply]
@aaidenduaban Are you using a `root` user while doing ssh from the host machine? If yes, just create a normal user in guest Linux and ssh with that user (not with root).  you are allowed to give sudo privilege to the user.

**Comment by mojcris** [comment]
First run `ifconfig` in your VM terminal. Usually there should be an item in the results called `enp0s1`. Find the ip address for in there and then ssh with the following command like you would do for any other server: 
```
ssh <USERNAME>@<IP_ADDRESS> 
``` 
If you can't ssh in using this method, make sure `ssh` is enabled in the VM. You can do that by running the following command: 
```
sudo systemctl enable ssh
``` 
Then try sshing again. Hopefully this works for you.

**Reply by sadmankiba** [reply]
This worked. I had to first install `ssh` server package in the VM.
```
sudo apt install openssh-server
sudo systemctl enable ssh
```

**Comment by rcorre** [comment]
If your host is on a VPN, the guest may not be able to acces WAN over `Shared Network` (see https://github.com/utmapp/UTM/discussions/3598). I was able to work around this by creating two separate networks:

- Bridged (default settings)
- Emulated VLAN with port forwarding 22->2022

Now the guest can access WAN over the first interface, and the host can SSH to the guest over the second. You may need to use `iproute` to ensure the guest uses the bridged interface for external connections.

**Comment by fmonday** [comment]
> make sure it's a valid IP that's on the same subnet as the host (the first 3 octets should be the same as your host)

I'm running it as "shared network" which is the only way I can ssh to it at all- but even then SSH only works inside the LAN (edit: and only on the host). From outside, with the router set to forward to the guest's IP at port 22 (from a DuckDNS URL), there's no response. I've tried everything and I think this bit about the subnet is where it's going wrong... the third octet isn't shared between the host and guest's IPs. I tried setting the "guest network" to, ex. 192.168.64.0/24 but then the VM won't even start.

**Comment by Teloshav** [comment]
## SSH INTO UTM VIRTUAL MACHINE - A GUIDE!

To save the sanity future me out as I'll definitely be back here again I'll show what I've done that's worked for my _Linux_ virtual machines. I hope this manages to help some other people out as well!

1. Open the settings for your virtual machine on UTM and navigate to the `Network` tab and change the following settings: 
- Network Mode: `Emulated VLAN`
- Emulated Network Card: `virtio-net-pci`

Here's a screenshot with the settings mentioned above (For future reference: taken September 2023):
![ssh_into_utm_vm 1](https://github.com/utmapp/UTM/assets/131397191/924c0167-9a83-4496-babd-0b0a810eb608)
<br>
2. Navigate to the `Port Forward` sub-menu and press `New...`. Use the following values:
- Protocol: `TCP`
- Guest Port: `22`
- Host Port: `22022` (Or any other appropriate value you choose, e.g. `5555`, `2022`, etc. Make sure this port isn't being used by anything else!)
- Other values: Blank
<br>
Here's the screenshots with the above settings added in UTM (For future reference: taken September 2023):

| `Port Forward` Sub-Menu | Adding new setting for port forwarding | The `Port Forward` Sub-Menu after adding the new settings |
| --- | --- | --- | 
| ![ssh_into_utm_vm 2](https://github.com/utmapp/UTM/assets/131397191/5986d375-b029-4b9f-8ed0-c00d280daa4a) | ![ssh_into_utm_vm 3](https://github.com/utmapp/UTM/assets/131397191/3c692910-7556-46b4-bbbc-ca81838af538) | ![ssh_into_utm_vm 4](https://github.com/utmapp/UTM/assets/131397191/3d9ca05f-fec3-4eef-8c4a-d8c70357e238) |


Press `Save` and you should be good to boot into your virtual machine! Setting the network to `VLAN` with the same Emulated Network Card has always been reliable when having internet problems with any of my virtual machines, Windows OR Linux! 😁 

### Steps 3-4 should be done in your virtual machine!!
3. Install `openssh-server` on your Linux **virtual machine** if it isn't already there. The command to do this varies from distribution to distribution but this is the command you use if you use `apt`:

```shell
sudo apt install openssh-server
```

4. Once that is installed you can use this command to start the service (this will happen whenever you boot up so there's nothing else you'll have to do when you start up you VM in the future. 😄):
```shell
systemctl enable ssh.service
# If the above command doesn't work try the following command:
systemctl enable sshd.service
```

5. On your **host machine** (MacOS), open `Terminal` or your terminal emulator of choice and enter one of the following commands:
```shell
ssh <username>@localhost -p <port number>
```
OR
```shell
ssh <username>@127.0.0.1 -p <port number>
```
`<username>` will be the username _inside the virtual machine_ that you want to log in as and `<port number>` is the "Host Port" you selected in `Step 2`, if you were following along with me then this would be `22022`.

6. You're in!! Time to party! 😁🥳🥳
---
With this I have been able to start using other architectures on my Apple ARM machine as using these operating systems _with_ their desktop interfaces was far too laggy and slow. Just make the VM headless and SSH into it! If you're familiar with a terminal as well this makes things so much easier as now you can copy and paste from the host machine without having to worry about [clipboard sharing](https://docs.getutm.app/guest-support/sharing/clipboard/).

Like I said earlier, these settings and whatnot have been successful for me and now I want to keep them here in the likely case I forget these settings and steps. I had been banging my head against this problem for almost 12 hours over the course of a few days just to get this working. I hope you all find this useful as it took me some time to type. :)

Have fun with your brand new ssh connection now!!

- Teloshav

P.S. I'll use a VPN when I am out and about as well, this shouldn't change anything but I'll be sure to update it if this isn't the case! [EDIT: I have been able to use this when connected to a VPN on my host machine.]

**Reply by Zicheng-Li** [reply]
GOAT!

**Reply by bartekpacia** [reply]
I had to run `systemctl enable sshd.service` instead of `systemctl enable ssh.service`. Thanks a lot for the guide!

**Reply by Teloshav** [reply]
> I had to run `systemctl enable sshd.service` instead of `systemctl enable ssh.service`. Thanks a lot for the guide!

I'll add this into the guide. :)

**Reply by yuyuUNny** [reply]
So kind and helpful!!!!

**Reply by tobiasschweizer** [reply]
That's great, I can ssh into the VM. However, I cannot get an internet connection from inside the VM when in a VPN.

**Reply by lucianthorr** [reply]
@tobiasschweizer If you're on ubuntu
You can just keep the Network on "Shared Network" and use `ssh ubuntu@ubuntu.local` to log in.
I noticed that hostname in the Ubuntu's Settings > Sharing > Remote Login, which needs to On.

**Reply by tobiasschweizer** [reply]
@lucianthorr Awesome, thanks! It is also possible to change the name in settings, but this requires a restart.
Note these are the settings of Ubuntu itself (not UTM's settings ;-))...

**Reply by radimurban** [reply]
Thanks a lot for this write up!

**Comment by startergo** [comment]
I can do a `Remote Desktop` to my Snow Leopard UTM and use the clipboard sharing. the shared folders are already enabled so I can access the guest and the guest can access the host using bridged network. This is better than ssh.

**Reply by daluu** [reply]
No UI lag with that approach? The guest is Show Leopard running in UTM with host being newer macOS on M1/M2 hardware?

**Reply by startergo** [reply]
Intel MBP15,1. There is no lag whatsoever. 
![image](https://github.com/utmapp/UTM/assets/7897244/fae191e9-12cd-4b10-a5c9-6e19f4c61e61)

You can run it on Silicon Mac as well:
https://github.com/royalgraphx/LegacyOSXKVM/blob/main/info/AS_SL.md

**Reply by adespoton** [reply]
Royal's image works great, but it doesn't work in UTM AArch64 -- just in vanilla QEMU and Intel UTM. I've updated it and streamlined it and saved it as a 3.64GB qcow2 image with sshd pre-enabled but no user accounts; I should be able to get it on archive.org for general consumption soon.

All that said, ssh is definitely snappy, and remote admin is even reasonable.

**Comment by achin-netskope** [comment]
Helpful page with step by step directions
https://arteen.linux.ucla.edu/ssh-into-utm-vm.html

**Reply by startergo** [reply]
You can do bridge connection. No need for NAT forwarding. And if you use the `virtIO driver` it is even better.
https://github.com/startergo/virtio-net-osx
<img width="1170" alt="image" src="https://github.com/utmapp/UTM/assets/7897244/cac1883a-a1a3-4095-9610-2cd35852d624">

**Comment by fred666333** [comment]
This works for me. Host is MacOS 15.5 and guest is CentOS Stream 9.

1. Create a bridged network in UTM.
2. Run these commands in guest: "sudo nmcli con add type bridge con-name br0 ifname br0" and "sudo nmcli con add type bridge-slave ifname enp0s1 master br0". More details here https://guix.gnu.org/cookbook/en/html_node/Network-bridge-for-QEMU.html.

To get the name of ethernet connection run "sudo nmcli connection" (Mine is enp0s1).

Now host can connect to guest.

Fred

**Comment by jaimergp** [comment]
When the guest is Windows, if both `ssh` and `ping` time out, you may need to enable some firewall rules and/or configure the network as Private:

```pwsh
# Allow SSH
New-NetFirewallRule -Name "OpenSSH-Server-In-TCP" -DisplayName "OpenSSH Server (Inbound)" -EventCmd "Allow" -Direction Inbound -Protocol TCP -LocalPort 22 -Action Allow
# Allow pings (for connection tests)
New-NetFirewallRule -Name "Allow-Ping-Inbound" -DisplayName "Allow ICMPv4 Echo Request" -Protocol ICMPv4 -IcmpType 8 -Action Allow -Enabled True
# Change network category
Get-NetConnectionProfile | Set-NetConnectionProfile -NetworkCategory Private
```
